src/USER-CONP/conp_matrix.cpp

```diff
@@ -0,0 +1,225 @@
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://lammps.sandia.gov/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing author: Naveen Michaud-Agrawal (Johns Hopkins U)
+     K-space terms added by Stan Moore (BYU)
+------------------------------------------------------------------------- */
+
+#include "conp_matrix.h"
+
+#include "atom.h"
+#include "comm.h"
+#include "error.h"
+#include "force.h"
+#include "group.h"
+#include "kspace.h"
+#include "math_const.h"
+#include "neigh_list.h"
+//#include "neigh_request.h"
+//#include "neighbor.h"
+#include "pair.h"
+
+using namespace LAMMPS_NS;
+using namespace MathConst;
+
+#define EWALD_F 1.12837917
+#define EWALD_P 0.3275911
+#define A1 0.254829592
+#define A2 -0.284496736
+#define A3 1.421413741
+#define A4 -1.453152027
+#define A5 1.061405429
+
+/* ---------------------------------------------------------------------- */
+
+ConpMatrix::ConpMatrix(LAMMPS *lmp, int igroup, double eta) : Pointers(lmp) {
+  this->igroup = igroup;
+  groupbit = group->bitmask[igroup];
+  ngroup = group->count(igroup);
+  array = new double *[ngroup];
+  for (bigint i = 0; i < ngroup; i++) array[i] = new double[ngroup];
+  this->eta = eta;
+}
+
+/* ---------------------------------------------------------------------- */
+
+ConpMatrix::~ConpMatrix() {
+  for (bigint i = 0; i < ngroup; i++) delete[] array[i];
+  delete[] array;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void ConpMatrix::setup(std::vector<int> tag_ids) {
+  // error if Kspace style does not compute coul/matrix interactions
+  if (force->kspace == nullptr)
+    error->all(FLERR, "No Kspace style defined for compute conp matrix");
+
+  // check if coul pair style is active, no need for single() since done
+  // explicitly
+  int itmp;
+  double *p_cutoff = (double *)force->pair->extract("cut_coul", itmp);
+  if (p_cutoff == nullptr)
+    error->all(FLERR, "conp matrix is incompatible with Pair style");
+  pair = force->pair;
+  cutsq = force->pair->cutsq;
+
+  conp_kspace = dynamic_cast<ConpKspace *>(force->kspace);
+  if (conp_kspace == nullptr)
+    error->all(FLERR, "Kspace does not implement ConpKspace");
+  g_ewald = force->kspace->g_ewald;
+
+  tag_to_iele = tag_ids;
```

Comment:
Make sure that atom IDs are consecutive (see "Atom IDs must be consecutive for velocity create loop all" in velocity.cpp)